### PR TITLE
A0-2061: Generate keys for merge relation

### PR DESCRIPTION
**Pull request by maciejzelaszczyk**
# Description

CLI for key generation, proof generation and verification for the `merge` relation in `aleph-node`. Can be tested using:

```bash
cargo run --release --features liminal -- snark-relation generate-keys merge --max-path-len 4
cargo run --release --features liminal -- snark-relation generate-proof --proving-key-file merge.groth16.pk.bytes merge --max-path-len 4 --token-id 1 --first-old-nullifier 19 --second-old-nullifier 29 --new-note "16873374675910556188,7520532372021931293,2252601191743571102,1137408349476470620" --merkle-root "7192045139731542264,1161130572820032929,11248828078021730941,7948961262277222795" --first-old-trapdoor 17 --second-old-trapdoor 23 --new-trapdoor 27 --new-nullifier 87 --first-merkle-path "15554272943220125889,94010064041599568,7999912732987454829,2345852616018906843:12424795876863646335,17547963776984128507,5386482043686570840,8102841056232622473" --second-merkle-path "16612482296235189599,15296027520069836296,4688001093087614490,6035486416775057912:7648513976096501199,6194936452710765692,2531672351598095972,5307215101199693779" --first-leaf-index 5 --second-leaf-index 6 --first-old-note "11425106013701053451,11840150414921744375,7525412728913434437,7906055471841466594" --second-old-note "17914626646753717819,12876986036806086814,4826170774187111240,739180176178530409" --first-old-token-amount 3 --second-old-token-amount 7 --new-token-amount 10
cargo run --release --features liminal -- snark-relation verify --verifying-key-file merge.groth16.vk.bytes --proof-file merge.groth16.proof.bytes --public-input-file merge.groth16.public_input.bytes
```

This scenario uses the same parameters as the `merge_proving_procedure_valid_circuit` test case.

## Type of change
- New feature (non-breaking change which adds functionality)

# Checklist:

- I have created new documentation

### CODE REVIEW

**Comment by pmikolajczyk41** [comment]
#958 accidentally removed some code from `cliain` and as the result these changes couldn't be tested. I have just pushed a fix - please test your changes now and add corresponding commands with data to the PR description, as in #819

File: bin/cliain/src/main.rs [hunk 1]
```diff
@@ -244,6 +247,84 @@ async fn main() -> anyhow::Result<()> {
...
+                proving_key_file,
+            } => generate_proof(relation, system, proving_key_file),
+            SnarkRelation::Verify {
+                verifying_key_file,
+                proof_file,
+                public_input_file,
+                system,
+            } => {
+                if verify_proof(verifying_key_file, proof_file, public_input_file, system) {
+                    println!("Proof is correct")
```
Comment (fbielejec) [inline bin/cliain/src/main.rs:322]: leftover?

File: bin/cliain/src/snark_relations/relations.rs [hunk 2]
```diff
@@ -201,7 +265,7 @@ impl ConstraintSynthesizer<CircuitField> for RelationArgs {
                 .generate_constraints(cs),
 
             RelationArgs::LinearEquation { a, x, b, y } => {
-                LinearEquationRelationWithFullInput::new(a, x, b, y).generate_constraints(cs)
+                LinearEquationRelationWithFullInput::new(a, b, y, x).generate_constraints(cs)
```
Comment (fbielejec) [inline bin/cliain/src/snark_relations/relations.rs:268]: was the order wrong before and you fixed it or did sth else happen here?

File: bin/cliain/Cargo.toml [hunk 3]
```diff
@@ -5,25 +5,28 @@ edition = "2021"
...
 log = "0.4"
-pallet-staking = { git = "https://github.com/Cardinal-Cryptography/substrate.git", branch = "aleph-v0.9.38" }
-primitives = { path = "../../primitives" }
 serde = { version = "1.0.137", features = ["derive"] }
 serde_json = "1.0.81"
-sp-core = { git = "https://github.com/Cardinal-Cryptography/substrate.git", branch = "aleph-v0.9.38", features = ["full_crypto"] }
 tokio = { version = "1.21.2", features = ["full"] }
 subxt = "0.25.0"
 
+contract-metadata = { git = "https://github.com/paritytech/cargo-contract.git", tag = "v1.4.0"}
```
Comment (fbielejec) [inline bin/cliain/Cargo.toml:22]: if I'm not mistaken deps were in alphabetic order, which probably isn't that bad to maintain as a convention?

File: bin/cliain/src/snark_relations/relations.rs (hunk 2 above)
Comment (pmikolajczyk41) [reply bin/cliain/src/snark_relations/relations.rs:268]: the order was wrong (the proper one changed when the relation was made with macro)

File: bin/cliain/Cargo.toml (hunk 3 above)
Comment (pmikolajczyk41) [reply bin/cliain/Cargo.toml:22]: some time ago it agreed that we should group deps:
 - crates.io
 - git
 - local

File: bin/cliain/src/main.rs (hunk 1 above)
Comment (maciejzelaszczyk) [reply bin/cliain/src/main.rs]: After discussion with @pmikolajczyk41, we would like to leave the info about the positive outcome of the verification.

**Review by fbielejec — APPROVED** [review]

**Review by pmikolajczyk41 — APPROVED** [review]
works for me now :muscle: however, please do not merge it to `snarkeling` now - after 
#986 is merged, we will remove `snarkeling` and merge this PR directly to `main`

**Comment by maciejzelaszczyk** [comment]
> commands from description will not work now - they need `--features liminal` + correct `aleph-client::aleph_zero.rs` module generated; I'll prepare patch in another PR

Great, have approved the patch.

**Pull request merged**